community/cypher/cypher-compiler-2.2/src/main/scala/org/neo4j/cypher/internal/compiler/v2_2/commands/expressions/KeysFunction.scala

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compiler.v2_2.commands.expressions
+
+import org.neo4j.cypher.internal.compiler.v2_2._
+import org.neo4j.cypher.internal.compiler.v2_2.executionplan.Effects
+import pipes.QueryState
+import symbols._
+import org.neo4j.cypher.internal.compiler.v2_2.spi.QueryContext
+import org.neo4j.graphdb.{Relationship, Node}
+
+case class KeysFunction(nodeOrRelationshipExpr: Expression) extends NullInNullOutExpression(nodeOrRelationshipExpr) {
+
+  override def compute(value: Any, m: ExecutionContext)(implicit state: QueryState) = value match {
+    case node: Node        => node.getId
+    case rel: Relationship => rel.getId
+    case x => throw new CypherTypeException("Expected `%s` to be a node or relationship, but it was ``".format(nodeOrRelationshipExpr, x.getClass.getSimpleName))
+  }
```

Comment:
Since you are extending NullInNullOutExpression, you should do the work in `compute()`. the node/relationship will be in `value`, so no need to execute `nodeOrRelationshipExpr` again.